docs/configuration_reference.rst
MediaMonksSonataMediaBundle Configuration Reference
===================================================

All available configuration options are listed below with their default values.

.. code-block:: yaml

    mediamonks_sonata_media:
        # service_id_of_a_flysystem_filesystem
        filesystem_private:
        # service_id_of_a_flysystem_filesystem
        filesystem_public:
        # redirect url to where files are stored
        redirect_url: 'https://url-to-where-your-files-are-stored/'
        # cache expire time which is set in the header of the redirect
        redirect_cache_ttl: 7776000
        # list of providers, by default all available providers are available
        providers:
            - 'MediaMonks\SonataMediaBundle\Provider\ImageProvider'
            - 'MediaMonks\SonataMediaBundle\Provider\FileProvider'
            - 'MediaMonks\SonataMediaBundle\Provider\YouTubeProvider'
            - 'MediaMonks\SonataMediaBundle\Provider\VimeoProvider'
            - 'MediaMonks\SonataMediaBundle\Provider\SoundCloudProvider'
        # an array with options supported by glide server
        glide: []
        # any parameter supported by glide image api
        default_image_parameters: []
        # any parameter supported by symfony image constraint
        image_constraints:
            minWidth: 100
            minHeight: 100
            maxWidth: 3000
            maxHeight: 3000
        # any parameter supported by symfony file constraint
        file_constraints:
            maxSize: '5M'
            extensions: ['pdf', 'csv', 'txt', 'docx']
        # path to an image file
        fallback_image: ~
        # path to local file system with read and write permission, defaults to sys_get_temp_dir() function
        tmp_path: ~
        # custom prefix for temporary files
        tmp_prefix: ~
        # route name used for image generation
        route_image: mediamonks_media_image
        # route name for downloads
        route_download: mediamonks_media_download
        # fqcn of the model
        model_class: App\Entity\Media
        # fqcn of the admin
        admin_class: MediaMonks\SonataMediaBundle\Admin\MediaAdmin
        # fqcn of the controller
        controller_class: MediaMonks\SonataMediaBundle\Controller\CRUDController


docs/2-configuring_the_bundle.rst
Step 2: Configuring the bundle
==============================


Providers
---------

A provider supports some kind of media. By default the bundle loads all available providers which are provided by
this bundle. You can add your own provider by tagging it as "sonata_media.provider".


File Systems
------------

This bundle requires you to have 2 configured filesystems with Flysystem. The "private" file system stores all
original uploads while the "public" file system stores all thumbnails. You can of course use the same file system
for both but this can be risky since you might allow uploads that can harm the machine the file is stored on (eg:
allow a php file to be uploaded and to be executed by the end user!). For the configuration of Flysystem we recommend
using the `Flysystem Bundle`_.

.. code-block:: yaml

    # app/config/config.yml
    mediamonks_sonata_media:
        filesystem_private: <service id of a flysystem file system>
        filesystem_public: <service id of a flysystem file system>

Redirect
--------

This bundle assumes you are using some kind of storage that is behind a CDN or a reverse proxy that supports cache
control headers. The benefit of this is that after an image is generated it does not hit our server again until it
expires. By default the ttl is set to 90 days.

.. code-block:: yaml

    # app/config/config.yml
    mediamonks_sonata_media:
        redirect_url: "<https://path.to.public.file.system/>"
        redirect_cache_ttl: 7776000


Image Parameters
----------------

While it is possible to set custom options while generating an image per media object you might want to always have some
default settings for the Glide Api:

.. code-block:: yaml

    # app/config/config.yml
    mediamonks_sonata_media:
        default_image_parameters:
            fit: "crop"


Image Constraints
-----------------

You can set image constraints to avoid bad quality media, below are the default settings:

.. code-block:: yaml

    # app/config/config.yml
    mediamonks_sonata_media:
        image_constraints:
            minWidth: 100
            minHeight: 100
            maxWidth: 5000
            maxHeight: 5000


File Constraints
----------------

Since this can be a really dangerous provider it is very restrictive by default. Please carefully consider to allow
file uploads of any file type. Below are the default settings:

.. code-block:: yaml

    # app/config/config.yml
    mediamonks_sonata_media:
        file_constraints:
            maxSize: '5M'
            extensions: ['pdf', 'csv', 'txt', 'docx']

Model Class
-----------

The bundle assumes that your entity is using FQCN ``App\Entity\Media``, if your entity is using a different
FQCN you can use the option ``model_class`` to set it.

.. code-block:: yaml

    # app/config/config.yml
    mediamonks_sonata_media:
        model_class: Acme\AppBundle\Entity\Media

Model Class
-----------

Override or extend the admin class with your own to modify behavior.

.. code-block:: yaml

    # app/config/config.yml
    mediamonks_sonata_media:
        admin_class: App\Admin\MediaAdmin


Controller Class
----------------

Override or extend the controller class with your own to modify behavior.

.. code-block:: yaml

    # app/config/config.yml
    mediamonks_sonata_media:
        controller_class: App\Controller\Admin\MediaCRUDController


.. _Flysystem Bundle: https://github.com/1up-lab/OneupFlysystemBundle


src/DependencyInjection/Configuration.php
<?php

namespace MediaMonks\SonataMediaBundle\DependencyInjection;

use MediaMonks\SonataMediaBundle\Admin\MediaAdmin;
use MediaMonks\SonataMediaBundle\Controller\CRUDController;
use MediaMonks\SonataMediaBundle\Entity\Media;
use MediaMonks\SonataMediaBundle\MediaMonksSonataMediaBundle;
use MediaMonks\SonataMediaBundle\Provider\FileProvider;
use MediaMonks\SonataMediaBundle\Provider\ImageProvider;
use MediaMonks\SonataMediaBundle\Provider\SoundCloudProvider;
use MediaMonks\SonataMediaBundle\Provider\VimeoProvider;
use MediaMonks\SonataMediaBundle\Provider\YouTubeProvider;
use Symfony\Component\Config\Definition\Builder\ArrayNodeDefinition;
use Symfony\Component\Config\Definition\Builder\TreeBuilder;
use Symfony\Component\Config\Definition\ConfigurationInterface;

final class Configuration implements ConfigurationInterface
{
    /**
     * Generates the configuration tree.
     *
     * @return TreeBuilder
     */
    public function getConfigTreeBuilder()
    {
        $treeBuilder = new TreeBuilder();
        $rootNode = $treeBuilder->root(MediaMonksSonataMediaBundle::BUNDLE_CONFIG_NAME);

        $this->addFilesystem($rootNode);
        $this->addRedirectUrl($rootNode);
        $this->addRedirectCacheTtl($rootNode);
        $this->addProviders($rootNode);
        $this->addGlideConfig($rootNode);
        $this->addDefaultImageParameters($rootNode);
        $this->addImageConstraints($rootNode);
        $this->addFileConstraints($rootNode);
        $this->addFallbackImage($rootNode);
        $this->addTmp($rootNode);
        $this->addTemplates($rootNode);
        $this->addRoutes($rootNode);
        $this->addClasses($rootNode);

        return $treeBuilder;
    }

    /**
     * @param ArrayNodeDefinition $node
     */
    private function addFilesystem(ArrayNodeDefinition $node)
    {
        $node->children()
            ->scalarNode('filesystem_private')
            ->end();

        $node->children()
            ->scalarNode('filesystem_public')
            ->defaultNull()
            ->end();
    }

    /**
     * @param ArrayNodeDefinition $node
     */
    private function addRedirectUrl(ArrayNodeDefinition $node)
    {
        $node->children()
            ->scalarNode('redirect_url')
            ->defaultNull()
            ->end();
    }

    /**
     * @param ArrayNodeDefinition $node
     */
    private function addRedirectCacheTtl(ArrayNodeDefinition $node)
    {
        $node->children()
            ->scalarNode('redirect_cache_ttl')
            ->defaultValue(60 * 60 * 24 * 90)// 90 days
            ->end();
    }

    /**
     * @param ArrayNodeDefinition $node
     */
    private function addProviders(ArrayNodeDefinition $node)
    {
        $node->children()
            ->arrayNode('providers')
            ->defaultValue(
                [
                    FileProvider::class,
                    ImageProvider::class,
                    YouTubeProvider::class,
                    VimeoProvider::class,
                    SoundCloudProvider::class
                ]
            )
            ->prototype('scalar')->end()
            ->end();
    }

    /**
     * @param ArrayNodeDefinition $node
     */
    private function addGlideConfig(ArrayNodeDefinition $node)
    {
        $node->children()
            ->arrayNode('glide')
            ->prototype('scalar')->end()
            ->end();
    }

    /**
     * @param ArrayNodeDefinition $node
     */
    private function addDefaultImageParameters(ArrayNodeDefinition $node)
    {
        $node->children()
            ->arrayNode('default_image_parameters')
            ->prototype('scalar')->end()
            ->end();
    }

    /**
     * @param ArrayNodeDefinition $node
     */
    private function addImageConstraints(ArrayNodeDefinition $node)
    {
        $node->children()
            ->variableNode('image_constraints')
            ->defaultValue([
                'minWidth' => 100,
                'minHeight' => 100,
                'maxWidth' => 3000,
                'maxHeight' => 3000
            ])
            ->end();
    }

    /**
     * @param ArrayNodeDefinition $node
     */
    private function addFileConstraints(ArrayNodeDefinition $node)
    {
        $node->children()
            ->variableNode('file_constraints')
            ->defaultValue([
                'maxSize' => '5M',
                'extensions' => ['pdf', 'csv', 'txt', 'docx']
            ])
            ->end();
    }

    /**
     * @param ArrayNodeDefinition $node
     */
    private function addFallbackImage(ArrayNodeDefinition $node)
    {
        $node->children()
            ->scalarNode('fallback_image')
            ->defaultNull()
            ->end();
    }

    /**
     * @param ArrayNodeDefinition $node
     */
    private function addTmp(ArrayNodeDefinition $node)
    {
        $node->children()
            ->scalarNode('tmp_path')
            ->defaultNull()
            ->end();

        $node->children()
            ->scalarNode('tmp_prefix')
            ->defaultNull()
            ->end();
    }

    /**
     * @param ArrayNodeDefinition $node
     */
    private function addTemplates(ArrayNodeDefinition $node)
    {
        $node->children()
            ->variableNode('templates')
            ->defaultValue([
                'form' => '@MediaMonksSonataMedia/Form/form.html.twig'
            ])
            ->end();
    }

    /**
     * @param ArrayNodeDefinition $node
     */
    private function addRoutes(ArrayNodeDefinition $node)
    {
        $node->children()
            ->scalarNode('route_image')
            ->defaultValue('mediamonks_media_image')
            ->end();

        $node->children()
            ->scalarNode('route_download')
            ->defaultValue('mediamonks_media_download')
            ->end();
    }

    /**
     * @param ArrayNodeDefinition $node
     */
    private function addClasses(ArrayNodeDefinition $node)
    {
        $node->children()
            ->scalarNode('model_class')
            ->defaultValue('App\Entity\Media')
            ->end();

        $node->children()
            ->scalarNode('admin_class')
            ->defaultValue(MediaAdmin::class)
            ->end();

        $node->children()
            ->scalarNode('controller_class')
            ->defaultValue(CRUDController::class)
            ->end();
    }
}
